File: sample/src/main/java/info/audio/rxrecorder/demo/MainActivity.java
package info.audio.rxrecorder.demo;

import android.Manifest;
import android.content.Intent;
import android.media.MediaRecorder;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.TextView;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

import info.audio.rxrecoder.ObservableAudioRecorder;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import library.minimize.com.chronometerpersist.ChronometerPersist;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;

@RuntimePermissions
public class MainActivity extends AppCompatActivity {
    private static final String FILE_NAME = "/sample.wav";
    private Disposable subscription;
    private ObservableAudioRecorder observableAudioRecorder;

    private Button record;
    private Button play;

    private ChronometerPersist chronometerPersist;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        TextView audioFormat = findViewById(R.id.textViewAudioFormat);
        Chronometer chronometer = findViewById(R.id.chronometer);

        chronometerPersist = ChronometerPersist.getInstance(chronometer, getSharedPreferences("MyPrefs", MODE_PRIVATE));
        final String filePath = Environment.getExternalStorageDirectory() + FILE_NAME;

        observableAudioRecorder = new ObservableAudioRecorder.Builder(MediaRecorder.AudioSource.CAMCORDER)
                .file(filePath)
                .build();

        audioFormat.setText(observableAudioRecorder.toString());

        play = findViewById(R.id.buttonPlay);
        record = findViewById(R.id.buttonRecord);
        record.setOnClickListener(view -> MainActivityPermissionsDispatcher.startRecordingWithCheck(MainActivity.this));

        subscription = Observable.create(observableAudioRecorder)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shorts -> {
                    try {
                        observableAudioRecorder.writeDataToFile(shorts);
                    } catch (IOException e) {
                        Log.e("Write", e.getMessage());
                    }
                }, throwable -> Log.e("Error", throwable.getMessage()));

        play.setVisibility(View.GONE);
        play.setOnClickListener(view -> playFile());
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        MainActivityPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (subscription != null) {
            observableAudioRecorder.stop();
            subscription.dispose();
        }
    }

    @NeedsPermission({Manifest.permission.RECORD_AUDIO, Manifest.permission.WRITE_EXTERNAL_STORAGE})
    void startRecording() {
        if (observableAudioRecorder.isRecording()) {
            try {
                observableAudioRecorder.stop();
                //helper method for closing the dataStream, also writes the wave header
                observableAudioRecorder.completeRecording();
                record.setText("Record");
            } catch (IOException e) {
                Log.e("Recorder", e.getMessage());
            }
            play.setVisibility(View.VISIBLE);
            chronometerPersist.stopChronometer();
        } else {
            try {
                observableAudioRecorder.start();
                record.setText("Stop");
            } catch (FileNotFoundException e) {
                Log.e("Recorder Error", e.getMessage());
            }
            chronometerPersist.startChronometer();
        }
    }

    private void playFile() {
        Intent intent = new Intent();
        intent.setAction(android.content.Intent.ACTION_VIEW);
        File file = new File(Environment.getExternalStorageDirectory(), FILE_NAME);
        Uri fileUri = FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID + ".audioprovider", file);
        intent.setDataAndType(fileUri, "audio/*");
        startActivity(intent);
    }
}

File: sample/src/main/java/info/audio/rxrecorder/demo/MainActivity.kt
package info.audio.rxrecorder.demo

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.media.MediaRecorder
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import info.audio.rxrecoder.ObservableAudioRecorder
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import library.minimize.com.chronometerpersist.ChronometerPersist
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.RuntimePermissions
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException

@RuntimePermissions
class MainActivity : AppCompatActivity() {
    private lateinit var filePath: String
    private lateinit var subscription: Disposable
    private lateinit var observableAudioRecorder: ObservableAudioRecorder

    private lateinit var chronometerPersist: ChronometerPersist

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        chronometerPersist = ChronometerPersist.getInstance(chronometer, getSharedPreferences("Prefs", Context.MODE_PRIVATE))
        filePath = this.externalCacheDir?.absolutePath + FILE_NAME

        observableAudioRecorder = ObservableAudioRecorder.Builder(MediaRecorder.AudioSource.CAMCORDER)
                .file(filePath)
                .findBestAudioRate()
                .build()

        textViewAudioFormat.text = observableAudioRecorder.toString()

        buttonRecord.setOnClickListener { toggleRecordingWithPermissionCheck() }

        subscription = Observable.create(observableAudioRecorder)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ shorts ->
                    try {
                        observableAudioRecorder.writeDataToFile(shorts)
                    } catch (e: IOException) {
                        Log.e("Write", "IOException")
                    }
                }) { throwable -> Log.e("Error", "${throwable.message}") }

        buttonPlay.visibility = View.GONE
        buttonPlay.setOnClickListener { playFile() }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }

    override fun onPause() {
        super.onPause()
        observableAudioRecorder.stop()
        subscription.dispose()
    }

    @SuppressLint("SetTextI18n")
    @NeedsPermission(Manifest.permission.RECORD_AUDIO, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    internal fun toggleRecording() {
        buttonPlay.visibility = View.VISIBLE
        if (observableAudioRecorder.isRecording()) {
            buttonPlay.isEnabled = true
            try {
                buttonRecord.text = "Record"
                observableAudioRecorder.stop()
                //helper method for closing the dataStream, also writes the wave header
                observableAudioRecorder.completeRecording()
            } catch (e: IOException) {
                Log.e("Recorder", "IOException")
            }

            buttonPlay.text = "Play"
            chronometerPersist.stopChronometer()
        } else {
            buttonPlay.isEnabled = false
            try {
                observableAudioRecorder.start()
                buttonRecord.text = "Stop"
            } catch (e: FileNotFoundException) {
                Log.e("Recorder Error", "FileNotFoundException")
            }

            chronometerPersist.startChronometer()
        }
    }

    private fun playFile() {
        val intent = Intent()
        intent.action = Intent.ACTION_VIEW
        val fileUri = FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID + ".audioprovider", File(filePath))
        intent.setDataAndType(fileUri, contentResolver.getType(fileUri))
        intent.putExtra(Intent.EXTRA_STREAM, fileUri)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        startActivity(Intent.createChooser(intent, "Play Sound File"))
    }

    companion object {
        private const val FILE_NAME = "/sample.wav"
    }
}
